### PR TITLE
feat: implement group Repository

### LINKED ISSUE #208: Create Group Repository
## Feature Title

Implement Group Repository

## Acceptance Criteria

- [ ] implement groupRepository

### PR DESCRIPTION
**Pull request by taniguchiiqqq**
## Overviews of implementation
add Group Repository

## Review points
I left some comments where I want you to review codes.

### CODE REVIEW

File: backend/main/src/Group/Repositories/GroupRepository.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	const groupId = groupIdOrErr.unwrap();
+	const name = groupnameOrErr.unwrap();
+
+	const groupOrError = Group.create(groupId, {
+		name: name.name,
+		userIds: groupData.userIds.map((userId) => {
+			const userIdOrError = UserId.create(userId);
+			if (userIdOrError.ok) return userIdOrError.unwrap();
+			throw new GroupRepositoryError(userIdOrError.error.message, []);
```
Comment (taniguchiiqqq) [inline backend/main/src/Group/Repositories/GroupRepository.ts]: Should I make a function converting the string to UserId for readability? or is this implementation fine?

File: backend/main/src/Group/Repositories/GroupRepository.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		}),
+		containerIds: groupData.containerIds.map((containerId) => {
+			const containerIdOrError = ContainerId.create(containerId);
+			if (containerIdOrError.ok) return containerIdOrError.unwrap();
+			throw new GroupRepositoryError(containerIdOrError.error.message, []);
```
Comment (taniguchiiqqq) [inline backend/main/src/Group/Repositories/GroupRepository.ts]: Should I make a function converting the string to ContainerId for readability? or is this implementation fine?

File: backend/main/src/Group/Repositories/GroupRepository.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+	async find(id: GroupId): Promise<Group | null>;
+	async find(id: ContainerId): Promise<Group | null>;
+	async find(id: GroupId | ContainerId): Promise<Group | null> {
+		const groupData: GroupWithRelationsPrimitiveData = {
+			groupId: "",
+			groupName: "",
+			userIds: [],
+			containerIds: [],
+		};
```
Comment (taniguchiiqqq) [inline backend/main/src/Group/Repositories/GroupRepository.ts]: I implement the find method by fulfilling the `groupData` object. But it's kinda hard to read, so let me know if you have a good idea that combines GroupDetailData , UserIds, and ContainerIds in a readable way.

Comment (hitohata) [reply backend/main/src/Group/Repositories/GroupRepository.ts]: You don't need to define the groupData. It is a mutable value, so this value leads to less readability.

File: backend/main/src/Group/Repositories/GroupRepository.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+			userIds: Group.userIds.map((userId) => userId.id),
+			containerIds: Group.containerIds.map((containerId) => containerId.id),
+		});
+	}
+
+	/**
+	 * Delete a Group
+	 * @param id - target Group id
+	 */
+	async delete(id: GroupId): Promise<undefined> {
```
Comment (hitohata) [inline backend/main/src/Group/Repositories/GroupRepository.ts:93]: How are you thinking of accomplish the deletion of a group? I mean, the group data has not only group data but also containers and the invitation link. So, to delete the group, these data also must be deleted.

https://genesis-tech-tribe.github.io/nishiki-documents/project-document/database#group

File: backend/main/src/Group/Repositories/GroupRepository.ts (hunk 1 above)
Comment (hitohata) [reply backend/main/src/Group/Repositories/GroupRepository.ts]: This string is used for logging, so you must write the error detail. Thinking the course of error, in this case, the user ID in the DB is incorrect. So for the case of that, we must write the log for checking errors.

see 26 line.
https://github.com/genesis-tech-tribe/nishiki-backend/blob/develop/backend/main/src/Shared/Layers/Controller/Controller.ts

File: backend/main/src/Group/Repositories/GroupRepository.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+			GroupDetailData.groupId,
+		);
+		const listOfUsersInGroup = await this.nishikiDbClient.listOfUsersInGroup(
+			GroupDetailData.groupId,
+		);
+		groupData.containerIds = listOfContainers;
+		groupData.userIds = listOfUsersInGroup.map((user) => user.userId);
+
+		return createGroupObject(groupData);
+	}
```
Comment (hitohata) [inline backend/main/src/Group/Repositories/GroupRepository.ts]:
```suggestion
	
	async find(id: GroupId | ContainerId): Promise<Group | null> {
		const groupDetailData =
			id instanceof GroupId
				? await this.nishikiDbClient.getGroup({
						groupId: id.id,
				  })
				: await this.nishikiDbClient.getGroup({
						containerId: id.id,
				  });

		if (!groupDetailData) return null;

		const groupId = groupDetailData.groupId;

		const [listOfContainers, listOfUsersInGroup] = await Promise.all([
			this.nishikiDbClient.listOfContainers(groupId),
			this.nishikiDbClient.listOfUsersInGroup(groupId),
		]);

		return createGroupObject({
			groupId: groupId,
			groupName: groupDetailData.groupName,
			userIds: listOfUsersInGroup.map((user) => user.userId),
			containerIds: listOfContainers,
		});
	}
```

**Review by hitohata — CHANGES_REQUESTED** [review]

File: backend/main/src/Group/Repositories/GroupRepository.ts (hunk 1 above)
Comment (taniguchiiqqq) [reply backend/main/src/Group/Repositories/GroupRepository.ts]: Thank you! I couldn't care about the logging. I've fixed them

File: backend/main/src/Group/Repositories/GroupRepository.ts (hunk 5 above)
Comment (taniguchiiqqq) [reply backend/main/src/Group/Repositories/GroupRepository.ts]: Thank you so much! Your implementation is way cleaner and more readable! I update it

File: backend/main/src/Group/Repositories/GroupRepository.ts (hunk 4 above)
Comment (taniguchiiqqq) [reply backend/main/src/Group/Repositories/GroupRepository.ts:93]: Ohh, sorry about that. That's true. Can I implement it after implementing deleteContainer method in NishikiDbClient?

**Comment by taniguchiiqqq** [comment]
@hitohata 
Thank you for your review! I've fixed them. I'm so sorry for the late update. Could you review it again?

File: backend/main/src/Group/Repositories/GroupRepository.ts (hunk 4 above)
Comment (hitohata) [reply backend/main/src/Group/Repositories/GroupRepository.ts:93]: OK. Let's keep it for now.

**Review by hitohata — APPROVED** [review]

**Pull request merged**